consumer/consumer.go

```
@@ -0,0 +1,172 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package consumer
+
+import (
+	"bufio"
+	"net"
+	"sync"
+
+	"github.com/m3db/m3msg/generated/proto/msgpb"
+	"github.com/m3db/m3msg/protocol/proto"
+)
+
+type listener struct {
+	net.Listener
+
+	opts    Options
+	msgPool *messagePool
+}
+
+// NewListener creates a consumer listener.
+func NewListener(addr string, opts Options) (Listener, error) {
+	if opts == nil {
+		opts = NewOptions()
+	}
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	mPool := newMessagePool(opts.MessagePoolOptions())
+	mPool.Init()
+	return &listener{
+		Listener: lis,
+		opts:     opts,
+		msgPool:  mPool,
+	}, nil
+}
+
+func (l *listener) Accept() (Consumer, error) {
+	conn, err := l.Listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+	return newConsumer(conn, l.msgPool, l.opts), nil
+}
+
+type consumer struct {
+	opts   Options
+	mPool  *messagePool
+	encdec proto.EncodeDecoder
+	rw     *bufio.ReadWriter
+	conn   net.Conn
+
+	ackLock sync.Mutex
+	ackPb   msgpb.Ack
+	canAck  bool
+}
+
+func newConsumer(
+	conn net.Conn,
+	mPool *messagePool,
+	opts Options,
+) *consumer {
+	r := bufio.NewReaderSize(conn, opts.ConnectionReadBufferSize())
+	w := bufio.NewWriterSize(conn, opts.ConnectionWriteBufferSize())
+	rw := bufio.NewReadWriter(r, w)
+	return &consumer{
+		opts:   opts,
+		mPool:  mPool,
+		encdec: proto.NewEncodeDecoder(rw, opts.EncodeDecoderOptions()),
+		rw:     rw,
+		conn:   conn,
+		canAck: true,
+	}
+}
+
+func (c *consumer) Message() (Message, error) {
+	m := c.mPool.Get()
+	m.reset(c)
+	if err := c.encdec.Decode(m); err != nil {
+		c.mPool.Put(m)
+		return nil, err
+	}
+	return m, nil
+}
+
+// This function could be called concurrently if messages are being
+// processed concurrently.
+func (c *consumer) ack(m msgpb.Metadata) {
+	c.ackLock.Lock()
+	if !c.canAck {
+		c.ackLock.Unlock()
+		return
```

Do we want to return an error here?

---

We don't, there is no way to handle that error, also consumers should just fire away the `ack`s and no longer care whether they are acked, in this case it's most likely the underlying connection is broken so the acks could not be sent back to producer anyway, the producer will retry those messages in a new consumer.

---

In that case would you name the function `tryAck` as the ack may succeed or fail and there's no way for the caller to know which?

---

sure sounds good